app.py

```diff
@@ -1,54 +1,128 @@
     
+    board_id = os.getenv('BOARD_ID')
     db_connectionstring = os.getenv('MONGO_CONNECTION_URL')
+    
 
     client = pymongo.MongoClient(
         db_connectionstring,
         tlsCAFile=certifi.where()
     )
     db = client.TodoListDB
     collection = db.todos
+
+    @login_manager.unauthorized_handler 
+    def unauthenticated():
+        uri = appClient.prepare_request_uri("https://github.com/login/oauth/authorize")
+        return redirect(uri)
+        
+    @login_manager.user_loader 
+    def load_user(user_id):
+        return User(user_id) 
+        # return None
+
+    @app.route('/login/callback', methods=['GET'])
+    def login_callback():
+        token_response = requests.post('https://github.com/login/oauth/access_token', data={
+                'client_id': client_id,
+                'client_secret': client_secret,
+                'code': request.args.get('code')
+            }, headers={'Accept': 'application/json'})
+
+        appClient.parse_request_body_response(token_response.text)
+
+        user_response = requests.get('https://api.github.com/user', headers={
+            'Accept': 'application/json', 
+            'Authorization': f"token {token_response.json()['access_token']}"
+            })
+
+        login_user(User(user_response.json()['login']))        
+        return redirect(url_for('index'))
     
+    def get_user_role():             
+        if (app.config.get('LOGIN_DISABLED') or current_user.user_id == admin_user):                        
+            return ROLE_WRITER
+        else:            
+            return ROLE_READER
+
     @app.route('/')
+    @login_required
     def index():
         items = mongoDB.get_items(collection, board_id)
-        item_view_model = ViewModel(items[0], items[1], items[2])
+        user_role = get_user_role()
+        item_view_model = ViewModel(items[0], items[1], items[2], user_role)
         return render_template('index.html', view_model=item_view_model)
 
     @app.route('/add', methods=['POST'])
+    @login_required
     def add():
         name = request.form.get('new_item_name')
         description = request.form.get('new_item_description')
-        mongoDB.create_item(collection, board_id, name, description)
-        return redirect(url_for('index'))
+        
+        user_role = get_user_role()
+        if (user_role == ROLE_WRITER):
+            mongoDB.create_item(collection, board_id, name, description)
+            return redirect(url_for('index'))
+        else:
+            return abort(403)
 
     @app.route('/start/<item_id>', methods=['POST'])
+    @login_required
     def start_item(item_id):
-        mongoDB.start_item(collection, board_id, item_id)
-        return redirect(url_for('index'))
-
+        user_role = get_user_role()
+        if (user_role == ROLE_WRITER):
+            mongoDB.start_item(collection, board_id, item_id)
+            return redirect(url_for('index'))
+        else:
+            return abort(403)
```

Comment:
Instead of duplicating this "if role == ... else abort" check for each function, you could try writing a function decorator.